python/ray/tune/integration/lightgbm.py

```diff
@@ -0,0 +1,176 @@
+    """LightGBM checkpoint callback
+
+    Saves checkpoints after each validation step.
+
+    Checkpoint are currently not registered if no ``tune.report()`` call
+    is made afterwards. Consider using ``TuneReportCheckpointCallback``
+    instead.
+
+    Args:
+        filename (str): Filename of the checkpoint within the checkpoint
+            directory. Defaults to "checkpoint".
+        frequency (int): How often to save checkpoints. Per default, a
+            checkpoint is saved every five iterations.
+
+    """
+    order = 19
+
+    def __init__(self, filename: str = "checkpoint", frequency: int = 5):
+        self._filename = filename
+        self._frequency = frequency
+
+    @staticmethod
+    def _create_checkpoint(model: Booster, epoch: int, filename: str,
+                           frequency: int):
+        if epoch % frequency > 0:
+            return
+        with tune.checkpoint_dir(step=epoch) as checkpoint_dir:
+            model.save_model(os.path.join(checkpoint_dir, filename))
+
+    def __call__(self, env: CallbackEnv) -> None:
+        self._create_checkpoint(env.model, env.iteration, self._filename,
+                                self._frequency)
+
+
+class TuneReportCheckpointCallback(TuneCallback):
+    """Creates a callback that reports metrics and checkpoints model.
+
+    Saves checkpoints after each validation step. Also reports metrics to Tune,
+    which is needed for checkpoint registration.
+
+    Args:
+        metrics (str|list|dict): Metrics to report to Tune. If this is a list,
+            each item describes the metric key reported to LightGBM,
+            and it will reported under the same name to Tune. If this is a
+            dict, each key will be the name reported to Tune and the respective
+            value will be the metric key reported to LightGBM.
+        filename (str): Filename of the checkpoint within the checkpoint
+            directory. Defaults to "checkpoint". If this is None,
+            all metrics will be reported to Tune under their default names as
+            obtained from LightGBM.
+        frequency (int): How often to save checkpoints. Per default, a
+            checkpoint is saved every five iterations.
+
+    Example:
+
+    .. code-block:: python
+
+        import lightgbm
+        from ray.tune.integration.lightgbm import (
+            TuneReportCheckpointCallback
+        )
+
+        config = {
+            # ...
+            "metric": ["binary_logloss", "binary_error"],
+        }
+
+        # Report only log loss to Tune after each validation epoch.
+        # Save model as `lightgbm.mdl`.
+        bst = lightgbm.train(
+            config,
+            train_set,
+            valid_sets=[test_set],
+            valid_names=["eval"],
+            verbose_eval=False,
+            callbacks=[TuneReportCheckpointCallback(
+                {"loss": "eval-binary_logloss"}, "lightgbm.mdl)])
+
+    """
+    order = 21
```

Comment:
what does `order` do?

Reply:
It's a LightGBM property. Callbacks with lower order will be ran first. It's a minor thing in this case but it's been set for completeness, as all built-in LGBM callbacks have it.